File: app/lib/types/restaurant.ts
import { Category } from './category';

export interface GooglePhoto {
  photo_reference: string | null;
  width: number | null;
  height: number | null;
  attribution_html: string | null;
  url: string | null;
}

export interface RestaurantListItem {
  id: string;
  slug: string;
  name: string;
  location_name: string;
  latitude: number | null;
  longitude: number | null;
  cover_image_url: string | null;
  computed_rating: number;
  review_count: number;
  category: Category | null;
}

export interface RestaurantDetail {
  id: string;
  slug: string;
  name: string;
  description: string | null;
  location_name: string;
  latitude: number | null;
  longitude: number | null;
  category_id: number | null;
  cover_image_url: string | null;
  computed_rating: number;
  review_count: number;
  created_by: string;
  created_at: string;
  updated_at: string;
  category: Category | null;
  google_place_id: string | null;
  website: string | null;
  phone_number: string | null;
  google_maps_url: string | null;
  price_level: number | null;
  opening_hours: string[] | null;
  // Fase B — Google Places enrichment
  google_rating: number | null;
  google_user_ratings_total: number | null;
  google_photos: GooglePhoto[] | null;
  editorial_summary: string | null;
  editorial_summary_lang: string | null;
  cuisine_types: string[] | null;
  google_cached_at: string | null;
  creator: {
    id: string;
    email: string;
    display_name: string;
    avatar_url: string | null;
    role: string;
    created_at: string;
    updated_at: string;
  } | null;
}

export interface CreateRestaurantRequest {
  slug: string;
  name: string;
  description?: string;
  location_name: string;
  latitude?: number;
  longitude?: number;
  cover_image_url?: string;
  category_id?: number;
  google_place_id?: string;
  website?: string;
  phone_number?: string;
  google_maps_url?: string;
  price_level?: number;
  opening_hours?: string[];
}

export type RatingDimensionKey = 'cleanliness' | 'ambiance' | 'service' | 'value' | 'food_quality';

export interface RestaurantRatingsResponse {
  restaurant_id: string;
  averages: Partial<Record<RatingDimensionKey, number>>;
  user_breakdown: Record<string, {
    dimension: RatingDimensionKey;
    score: number;
    user_display_name: string;
  }[]>;
}

// ----- Aggregation types (perfil-page endpoints) -----

export interface ProsConsAggregateItem {
  text: string;
  count: number;
}

export interface DimensionAggregate {
  average: number | null;
  count: number;
}

export interface RestaurantAggregates {
  pros_top: ProsConsAggregateItem[];
  cons_top: ProsConsAggregateItem[];
  dimension_averages: Partial<Record<RatingDimensionKey, DimensionAggregate>>;
  photos_count: number;
  dishes_count: number;
  reviews_count: number;
}

export interface RestaurantPhoto {
  id: string;
  url: string;
  alt_text: string | null;
  taken_at: string;
  dish_id: string;
  dish_name: string;
  review_id: string | null;
  user_id: string;
  user_handle: string | null;
  user_display_name: string;
}

export interface RestaurantPhotosResponse {
  items: RestaurantPhoto[];
  next_cursor: string | null;
}

export interface DiaryVisitor {
  id: string;
  handle: string | null;
  display_name: string;
  avatar_url: string | null;
}

export interface MostOrderedDish {
  id: string;
  name: string;
  review_count: number;
}

export interface DiaryStats {
  unique_visitors: number;
  visits_total: number;
  visits_last_7d: number;
  most_ordered_dish: MostOrderedDish | null;
  recent_visitors: DiaryVisitor[];
}

export interface SignatureDish {
  id: string;
  name: string;
  cover_image_url: string | null;
  computed_rating: number;
  review_count: number;
  best_quote: string | null;
  best_quote_author: string | null;
}

export interface SignatureDishesResponse {
  items: SignatureDish[];
}

export interface NearbyRestaurantItem {
  id: string;
  slug: string;
  name: string;
  location_name: string;
  cover_image_url: string | null;
  google_photo_url: string | null;
  computed_rating: number;
  review_count: number;
  category: Category | null;
  distance_km: number;
}

export interface NearbyRestaurantsResponse {
  items: NearbyRestaurantItem[];
}

export interface OpenStatusInfo {
  isOpen: boolean;
  closesAt?: string;
  opensAt?: string;
  todayLabel?: string;
  hasHours: boolean;
}

/** @deprecated Use RestaurantListItem instead */
export interface Restaurant {
  id: number;
  name: string;
  position: {
    lat: number;
    lng: number;
  };
  image: string;
  location: string;
  rating: number;
  description: string;
  reviewCount: number;
  category: string;
}

File: backend/app/schemas/restaurant.py
import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, Field

from app.models.restaurant import ProsConsType, RatingDimension
from app.schemas.category import CategoryResponse
from app.schemas.user import UserResponse


class RestaurantCreate(BaseModel):
    slug: str = Field(max_length=200)
    name: str = Field(max_length=200)
    description: str | None = None
    location_name: str = Field(max_length=300)
    latitude: Decimal | None = None
    longitude: Decimal | None = None
    category_id: int
    cover_image_url: str | None = None


class RestaurantUpdate(BaseModel):
    slug: str | None = Field(None, max_length=200)
    name: str | None = Field(None, max_length=200)
    description: str | None = None
    location_name: str | None = Field(None, max_length=300)
    latitude: Decimal | None = None
    longitude: Decimal | None = None
    category_id: int | None = None
    cover_image_url: str | None = None


class RestaurantResponse(BaseModel):
    id: uuid.UUID
    slug: str
    name: str
    description: str | None
    location_name: str
    latitude: Decimal | None
    longitude: Decimal | None
    category_id: int
    cover_image_url: str | None
    computed_rating: Decimal
    review_count: int
    created_by: uuid.UUID
    created_at: datetime
    updated_at: datetime
    category: CategoryResponse | None = None
    creator: UserResponse | None = None

    model_config = {"from_attributes": True}


class RestaurantListResponse(BaseModel):
    id: uuid.UUID
    slug: str
    name: str
    location_name: str
    cover_image_url: str | None
    computed_rating: Decimal
    review_count: int
    category: CategoryResponse | None = None

    model_config = {"from_attributes": True}


class RatingDimensionCreate(BaseModel):
    dimension: RatingDimension
    score: Decimal = Field(ge=1, le=5)


class RatingDimensionResponse(BaseModel):
    id: int
    restaurant_id: uuid.UUID
    user_id: uuid.UUID
    dimension: RatingDimension
    score: Decimal

    model_config = {"from_attributes": True}


class ProsConsCreate(BaseModel):
    type: ProsConsType
    text: str = Field(max_length=500)


class ProsConsResponse(BaseModel):
    id: int
    restaurant_id: uuid.UUID
    user_id: uuid.UUID
    type: ProsConsType
    text: str

    model_config = {"from_attributes": True}


class VisitDiaryEntryCreate(BaseModel):
    visit_date: date
    diary_text: str


class VisitDiaryEntryResponse(BaseModel):
    id: int
    restaurant_id: uuid.UUID
    visit_date: date
    diary_text: str
    created_by: uuid.UUID
    created_at: datetime

    model_config = {"from_attributes": True}
